source/extensions/tracers/datadog/datadog_tracer_impl.cc

```diff
@@ -0,0 +1,129 @@
+
+#include "common/common/enum_to_int.h"
+#include "common/common/fmt.h"
+#include "common/common/utility.h"
+#include "common/http/headers.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+#include "common/tracing/http_tracer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+
+Driver::TlsTracer::TlsTracer(const std::shared_ptr<opentracing::Tracer>& tracer,
+                             TraceReporterPtr&& reporter, Driver& driver)
+    : tracer_(tracer), reporter_(std::move(reporter)), driver_(driver) {}
+
+Driver::Driver(const envoy::config::trace::v2::DatadogConfig& datadog_config,
+               Upstream::ClusterManager& cluster_manager, Stats::Store& stats,
+               ThreadLocal::SlotAllocator& tls, Runtime::Loader& runtime)
+    : OpenTracingDriver{stats},
+      cm_(cluster_manager), tracer_stats_{DATADOG_TRACER_STATS(
+                                POOL_COUNTER_PREFIX(stats, "tracing.datadog."))},
+      tls_(tls.allocateSlot()), runtime_(runtime) {
+
+  Upstream::ThreadLocalCluster* cluster = cm_.get(datadog_config.collector_cluster());
+  if (!cluster) {
+    throw EnvoyException(fmt::format("{} collector cluster is not defined on cluster manager level",
+                                     datadog_config.collector_cluster()));
+  }
+  cluster_ = cluster->info();
+
+  // Default tracer options.
+  tracer_options_.operation_name_override = "envoy.proxy";
+  tracer_options_.service = "envoy";
+  tracer_options_.priority_sampling = true;
+
+  // Configuration overrides for tracer options.
+  if (!datadog_config.service_name().empty()) {
+    tracer_options_.service = datadog_config.service_name();
+  }
+  if (!datadog_config.priority_sampling()) {
+    tracer_options_.priority_sampling = false;
+  }
+
+  tls_->set([this](Event::Dispatcher& dispatcher) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    auto tp = datadog::opentracing::makeTracerAndEncoder(tracer_options_);
+    auto tracer = std::get<0>(tp);
+    auto encoder = std::get<1>(tp);
+    TraceReporterPtr reporter(new TraceReporter(encoder, *this, dispatcher));
+    return ThreadLocal::ThreadLocalObjectSharedPtr{
+        new TlsTracer(tracer, std::move(reporter), *this)};
+  });
+}
+
+opentracing::Tracer& Driver::tracer() { return *tls_->getTyped<TlsTracer>().tracer_; }
+
+TraceReporter::TraceReporter(TraceEncoderPtr encoder, Driver& driver, Event::Dispatcher& dispatcher)
+    : driver_(driver), encoder_(encoder) {
+  flush_timer_ = dispatcher.createTimer([this]() -> void {
+    driver_.tracerStats().timer_flushed_.inc();
+    flushTraces();
+    enableTimer();
+  });
+
+  enableTimer();
+}
+
+void TraceReporter::enableTimer() {
+  const uint64_t flush_interval =
+      driver_.runtime().snapshot().getInteger("tracing.datadog.flush_interval_ms", 1000U);
+  flush_timer_->enableTimer(std::chrono::milliseconds(flush_interval));
+}
+
+void TraceReporter::flushTraces() {
```

Comment:
I'm guessing that almost all this code was basically copied from the LS implementation? How much of it is different? Is it worth sharing the code in any way? cc @rnburn

Reply:
Correct, they were used as a reference, and where possible, used verbatim so that the result would be consistent.
If there's room to refactor things, that's great.
Flush timers and flush mechanisms are probably a good case for that.
Not sure about other bits.

Reply:
It's up to you but any de-dup would be appreciated. This is going to come up again when someone comes in and adds the OpenCensus tracer.

Reply:
Not done, but OK doing a followup change to de-dup things, and any suggestions from @rnburn